gapic/src/main/com/google/gapid/widgets/TabComposite.java

```diff
@@ -913,12 +952,49 @@ public void addTabToLargestFolder(TabInfo tab) {
       newTab(tab);
     }
 
+    @Override
+    public TabInfo pinTab(Object id) {
+      for (Tab tab : tabs) {
+        if (Objects.equals(tab.info.id, id)) {
+          if (tab.content.isPinnable()) {
+            tab.content.pin();
+            // assign new id for pinned tab
+            TabInfo oldInfo = tab.info;
+            tab.info = new TabInfo(tab.content, oldInfo.view, oldInfo.label, oldInfo.contentFactory);
+            return oldInfo;
+          } else {
+            return null;
+          }
+        }
+      }
+      return null;
+    }
+
+    @Override
+    public void disposePinnedTabs() {
+      int index = 0;
+      while (index < tabs.size()) {
+        Tab tab = tabs.get(index);
+        if (tab.info.id == tab.content) {
```

Comment:
This predicate seems a bit unintuitive to me to check whether this tab is a pinned tab. I'm wondering is there a reason that `tab.content.isPinned()` doesn't fit in here?

Reply:
Thanks, good point! I have added the function Tab.isPinned() to make it more readable.
The class TabComposite does not control if TabContent.isPinned() reliably always returns true when it is pinned, so there is a danger of a crash if it is not correctly implemented. However TabComposite.Folder does control the value of tab.info.id when it pins a tab, so it is more safe to use that value here.